src/app/js/formats/streamgraph/utils.js

```diff
@@ -0,0 +1,175 @@
+  }
+  const toHex = x => {
+    const hex = Math.round(x * 255).toString(16);
+    return hex.length === 1 ? "0" + hex : hex;
+  };
+  return `#${toHex(r)}${toHex(g)}${toHex(b)}`;
+}
+
+export function transformDataIntoMapArray(formatData) {
+  let dateMin = -42;
+  let dateMax = -42;
+  const valuesObjectsArray = [];
+
+  if (formatData) {
+    for (let i = 0; i < formatData.length; i++) {
+      let elem = formatData[i];
+      if (elem) {
+        // source: "2010"
+        // target: "Pétrogenèse de roches basaltiques"
+        // weight: 0.9013102637325718
+
+        if (elem.source && elem.weight) {
+          let currentItem = undefined;
+
+          for (let tmpElem of valuesObjectsArray) {
+            if (tmpElem.name === elem.target) {
+              currentItem = tmpElem;
+            }
+          }
+
+          if (currentItem === undefined) {
+            currentItem = {
+              name: elem.target,
+              values: []
+            };
+            valuesObjectsArray.push(currentItem);
+          }
+
+          currentItem.values.push({
+            date: new Date(elem.source),
+            value: elem.weight
+          });
+
+          if (parseInt(elem.source, 10) < dateMin || dateMin === -42) {
+            dateMin = parseInt(elem.source, 10);
+          }
+          if (parseInt(elem.source, 10) > dateMax || dateMax === -42) {
+            dateMax = parseInt(elem.source, 10);
+          }
+        }
+      }
+    }
+    // display the datas at the center when only one x value
+    if (dateMin == dateMax) {
+      dateMin--;
+      dateMax++;
+    }
+  }
+
+  const namesList = valuesObjectsArray.map(value => value.name);
+
+  let currentDate = dateMin;
+  const valuesArray = [];
+
+  while (currentDate <= dateMax) {
+    let tmpName = [];
+    let newElem = {
+      date: new Date(String(currentDate))
+    };
+
+    for (let element of valuesObjectsArray) {
+      // loop which add each values for the good year
+      for (let dateValue of element.values) {
+        if (dateValue.date.getFullYear() == currentDate) {
+          newElem[element.name] = dateValue.value;
+          tmpName.push(element.name);
+        }
+      }
+
+      // loop which add 0 value to the missing keys
+      let resMissingNameList = differenceBy(namesList, tmpName);
+      for (let elemToAdd of resMissingNameList) {
+        newElem[elemToAdd] = 0;
+      }
+    }
+
+    valuesArray.push(newElem);
+    currentDate++;
+  }
+
+  return { valuesObjectsArray, valuesArray, dateMin, dateMax, namesList };
+}
+
+export function getMinMaxValue(stackedData) {
+  let minValue = 0;
+  let maxValue = 0;
+
+  if (stackedData) {
+    for (let element of stackedData) {
+      for (let value of element) {
+        if (minValue > value[0]) {
+          minValue = value[0];
+        }
+        if (maxValue < value[1]) {
+          maxValue = value[1];
+        }
+      }
+    }
+  }
+
+  return { minValue, maxValue };
+}
+
+export function cutStr(str) {
```

Comment:
Let me help you with this function.

```js
// Naming is important: please don't use shurtnames (like str)
// It's always better to use full meaningful names for variables, it helps the reader a lot
export function limitWords = (str, maxWordNumber = 3) => {
    const words = str.split(' ');
    const sentence = words.filter((word, index) => index <= maxWordNumber).join(' ');

    return words.length > maxWordNumber
        ? `${sentence} ...`
        : sentence;
};
```

Reply:
Your solution doesn't produce the same result.

Reply:
I even haven't understood your code, then.
The goal isn't to get the first three words and add `...` if there is a rest?

Reply:
You missed the if conditions.. look [here](https://repl.it/@taspai/BoldAntiqueMemoryallocator)

Reply:
I get a 404 Not Found :/